File: packages/python/halt/core/stats.py
"""In-process statistics collector for rate-limit observability.

Implements the ``TelemetryHooks`` protocol, so you plug it into a RateLimiter
(and optionally QuotaManager / PenaltyManager) and it aggregates everything in
memory. Call ``snapshot()`` to expose a ``/halt/stats`` endpoint or log periodically.

It is per-process by design — for a fleet, run it on each instance and roll up
across instances with the OpenTelemetry adapter (see halt.observability.otel).
Memory is bounded: the tracked-key table is capped and the smallest counters are
evicted.

    stats = StatsCollector()
    limiter = RateLimiter(store=store, policy=policy, telemetry=stats)
    # ... later ...
    return stats.snapshot()
"""

from typing import Any, Dict, Optional

from halt.core.decision import Decision


class StatsCollector:
    """Per-process aggregator of rate-limit telemetry."""

    def __init__(self, top_n: int = 20, max_tracked_keys: int = 10000) -> None:
        """Initialize the collector.

        Args:
            top_n: How many top limited keys ``snapshot()`` returns.
            max_tracked_keys: Max distinct keys tracked before eviction.
        """
        self.top_n = top_n
        self.max_tracked_keys = max_tracked_keys

        self.allowed_total = 0
        self.blocked_total = 0
        self._by_policy: Dict[str, Dict[str, int]] = {}
        self._by_endpoint: Dict[str, Dict[str, float]] = {}
        self._limited_keys: Dict[str, int] = {}
        self._cost_by_plan: Dict[str, float] = {}
        self.quota_exceeded = 0
        self.penalties_applied = 0
        self.violations = 0

    # ----- telemetry hooks -------------------------------------------------

    def on_check(self, key: str, decision: Decision, metadata: Optional[dict] = None) -> None:
        """No-op (counts are recorded in on_allowed / on_blocked)."""

    def on_allowed(self, key: str, decision: Decision, metadata: Optional[dict] = None) -> None:
        self.allowed_total += 1
        meta = metadata or {}
        cost = float(meta.get("cost", 1))
        self._policy_tally(meta)["allowed"] += 1

        endpoint = meta.get("endpoint")
        if endpoint:
            e = self._endpoint_tally(endpoint)
            e["allowed"] += 1
            e["cost"] += cost

        plan = meta.get("plan") or meta.get("policy") or "unknown"
        self._cost_by_plan[plan] = self._cost_by_plan.get(plan, 0) + cost

    def on_blocked(self, key: str, decision: Decision, metadata: Optional[dict] = None) -> None:
        self.blocked_total += 1
        meta = metadata or {}
        self._policy_tally(meta)["blocked"] += 1

        endpoint = meta.get("endpoint")
        if endpoint:
            self._endpoint_tally(endpoint)["blocked"] += 1

        if key:
            self._track_limited_key(key)

    def on_quota_check(self, identifier: str, quota: Any, allowed: bool) -> None:
        """No-op (exceedances are recorded in on_quota_exceeded)."""

    def on_quota_exceeded(self, identifier: str, quota: Any) -> None:
        self.quota_exceeded += 1

    def on_penalty_applied(self, identifier: str, penalty: Any) -> None:
        self.penalties_applied += 1

    def on_violation(self, identifier: str, penalty: Any, severity: float) -> None:
        self.violations += 1

    # ----- output ----------------------------------------------------------

    def snapshot(self) -> dict:
        """Point-in-time view of the aggregated stats."""
        top = sorted(self._limited_keys.items(), key=lambda kv: kv[1], reverse=True)
        top_limited_keys = [{"key": k, "blocked": v} for k, v in top[: self.top_n]]

        return {
            "allowed_total": self.allowed_total,
            "blocked_total": self.blocked_total,
            "by_policy": {k: dict(v) for k, v in self._by_policy.items()},
            "by_endpoint": {k: dict(v) for k, v in self._by_endpoint.items()},
            "top_limited_keys": top_limited_keys,
            "cost_by_plan": dict(self._cost_by_plan),
            "quota_exceeded": self.quota_exceeded,
            "penalties_applied": self.penalties_applied,
            "violations": self.violations,
            "tracked_keys": len(self._limited_keys),
        }

    def reset(self) -> None:
        """Clear all counters (e.g. after exporting)."""
        self.allowed_total = 0
        self.blocked_total = 0
        self._by_policy.clear()
        self._by_endpoint.clear()
        self._limited_keys.clear()
        self._cost_by_plan.clear()
        self.quota_exceeded = 0
        self.penalties_applied = 0
        self.violations = 0

    # ----- helpers ---------------------------------------------------------

    def _policy_tally(self, meta: dict) -> Dict[str, int]:
        policy = meta.get("policy") or "unknown"
        t = self._by_policy.get(policy)
        if t is None:
            t = {"allowed": 0, "blocked": 0}
            self._by_policy[policy] = t
        return t

    def _endpoint_tally(self, endpoint: str) -> Dict[str, float]:
        e = self._by_endpoint.get(endpoint)
        if e is None:
            e = {"allowed": 0, "blocked": 0, "cost": 0}
            self._by_endpoint[endpoint] = e
        return e

    def _track_limited_key(self, key: str) -> None:
        if key in self._limited_keys:
            self._limited_keys[key] += 1
            return
        if len(self._limited_keys) >= self.max_tracked_keys:
            self._evict_smallest()
        self._limited_keys[key] = 1

    def _evict_smallest(self) -> None:
        if not self._limited_keys:
            return
        min_key = min(self._limited_keys, key=self._limited_keys.get)
        del self._limited_keys[min_key]

File: packages/typescript/src/core/stats.ts
/**
 * In-process statistics collector for rate-limit observability.
 *
 * Implements `TelemetryHooks`, so you plug it into a RateLimiter (and optionally
 * QuotaManager / PenaltyManager) and it aggregates everything in memory. Call
 * `snapshot()` to expose a `/halt/stats` endpoint or log periodically.
 *
 * It is per-process by design — for a fleet, run it on each instance and roll up
 * across instances with the OpenTelemetry adapter (see observability/otel). Memory
 * is bounded: the tracked-key table is capped and the smallest counters are evicted.
 *
 *   const stats = new StatsCollector();
 *   const limiter = new RateLimiter({ store, policy, telemetry: stats });
 *   app.get('/halt/stats', (_req, res) => res.json(stats.snapshot()));
 */

import { Decision } from './decision';
import { TelemetryHooks } from './telemetry';
import { Quota } from './quota';
import { Penalty } from './penalty';

export interface StatsCollectorOptions {
    /** How many top limited keys `snapshot()` returns (default 20). */
    topN?: number;
    /** Max distinct keys tracked for the top-N table before eviction (default 10000). */
    maxTrackedKeys?: number;
}

interface Tally {
    allowed: number;
    blocked: number;
}

interface EndpointTally extends Tally {
    cost: number;
}

export interface StatsSnapshot {
    allowedTotal: number;
    blockedTotal: number;
    /** Per-policy allowed/blocked counts. */
    byPolicy: Record<string, Tally>;
    /** Per-endpoint allowed/blocked counts and consumed cost. */
    byEndpoint: Record<string, EndpointTally>;
    /** Highest-blocked keys (users / API keys / IPs), descending. */
    topLimitedKeys: Array<{ key: string; blocked: number }>;
    /** Consumed cost per plan (allowed requests only). */
    costByPlan: Record<string, number>;
    /** Count of quota-exceeded events (from QuotaManager telemetry). */
    quotaExceeded: number;
    /** Count of penalties applied (from PenaltyManager telemetry). */
    penaltiesApplied: number;
    /** Count of recorded violations (from PenaltyManager telemetry). */
    violations: number;
    /** Number of distinct keys currently tracked for top-N. */
    trackedKeys: number;
}

export class StatsCollector implements TelemetryHooks {
    private topN: number;
    private maxTrackedKeys: number;

    private allowedTotal = 0;
    private blockedTotal = 0;
    private byPolicy = new Map<string, Tally>();
    private byEndpoint = new Map<string, EndpointTally>();
    private limitedKeys = new Map<string, number>();
    private costByPlan = new Map<string, number>();
    private quotaExceededCount = 0;
    private penaltiesAppliedCount = 0;
    private violationsCount = 0;

    constructor(options: StatsCollectorOptions = {}) {
        this.topN = options.topN ?? 20;
        this.maxTrackedKeys = options.maxTrackedKeys ?? 10000;
    }

    onAllowed(_key: string, _decision: Decision, metadata?: Record<string, any>): void {
        this.allowedTotal++;
        const cost = Number(metadata?.cost ?? 1);
        this.policyTally(metadata).allowed++;

        const endpoint = metadata?.endpoint;
        if (endpoint) {
            const e = this.endpointTally(endpoint);
            e.allowed++;
            e.cost += cost;
        }

        const plan = metadata?.plan ?? metadata?.policy ?? 'unknown';
        this.costByPlan.set(plan, (this.costByPlan.get(plan) ?? 0) + cost);
    }

    onBlocked(key: string, _decision: Decision, metadata?: Record<string, any>): void {
        this.blockedTotal++;
        this.policyTally(metadata).blocked++;

        const endpoint = metadata?.endpoint;
        if (endpoint) {
            this.endpointTally(endpoint).blocked++;
        }

        if (key) this.trackLimitedKey(key);
    }

    onQuotaExceeded(_identifier: string, _quota: Quota): void {
        this.quotaExceededCount++;
    }

    onPenaltyApplied(_identifier: string, _penalty: Penalty): void {
        this.penaltiesAppliedCount++;
    }

    onViolation(_identifier: string, _penalty: Penalty, _severity: number): void {
        this.violationsCount++;
    }

    /** Point-in-time view of the aggregated stats. */
    snapshot(): StatsSnapshot {
        const topLimitedKeys = [...this.limitedKeys.entries()]
            .sort((a, b) => b[1] - a[1])
            .slice(0, this.topN)
            .map(([key, blocked]) => ({ key, blocked }));

        return {
            allowedTotal: this.allowedTotal,
            blockedTotal: this.blockedTotal,
            byPolicy: mapToObject(this.byPolicy),
            byEndpoint: mapToObject(this.byEndpoint),
            topLimitedKeys,
            costByPlan: mapToObject(this.costByPlan),
            quotaExceeded: this.quotaExceededCount,
            penaltiesApplied: this.penaltiesAppliedCount,
            violations: this.violationsCount,
            trackedKeys: this.limitedKeys.size,
        };
    }

    /** Clear all counters (e.g. after exporting). */
    reset(): void {
        this.allowedTotal = 0;
        this.blockedTotal = 0;
        this.byPolicy.clear();
        this.byEndpoint.clear();
        this.limitedKeys.clear();
        this.costByPlan.clear();
        this.quotaExceededCount = 0;
        this.penaltiesAppliedCount = 0;
        this.violationsCount = 0;
    }

    private policyTally(metadata?: Record<string, any>): Tally {
        const policy = metadata?.policy ?? 'unknown';
        let t = this.byPolicy.get(policy);
        if (!t) {
            t = { allowed: 0, blocked: 0 };
            this.byPolicy.set(policy, t);
        }
        return t;
    }

    private endpointTally(endpoint: string): EndpointTally {
        let e = this.byEndpoint.get(endpoint);
        if (!e) {
            e = { allowed: 0, blocked: 0, cost: 0 };
            this.byEndpoint.set(endpoint, e);
        }
        return e;
    }

    private trackLimitedKey(key: string): void {
        const current = this.limitedKeys.get(key);
        if (current !== undefined) {
            this.limitedKeys.set(key, current + 1);
            return;
        }
        if (this.limitedKeys.size >= this.maxTrackedKeys) {
            this.evictSmallest();
        }
        this.limitedKeys.set(key, 1);
    }

    /** Evict the lowest-count tracked key to keep memory bounded. */
    private evictSmallest(): void {
        let minKey: string | undefined;
        let minVal = Infinity;
        for (const [k, v] of this.limitedKeys) {
            if (v < minVal) {
                minVal = v;
                minKey = k;
            }
        }
        if (minKey !== undefined) this.limitedKeys.delete(minKey);
    }
}

function mapToObject<V>(map: Map<string, V>): Record<string, V> {
    const out: Record<string, V> = {};
    for (const [k, v] of map) out[k] = v;
    return out;
}
